Fix some minor scheduler issues with groups

kgaillot opened the pull request:
The scheduler group code is full of corner cases that are not correctly handled. This is the start of a project to clean them up.

It starts by adding four new regression tests, one of which indicates a problem that is not corrected in this PR but hopefully will be separately.

kgaillot commented:
@nrwahl2 , can you review this when you get a chance? Thanks ...

nrwahl2 commented on lib/pacemaker/pcmk_sched_colocation.c:
```diff
@@ -1265,6 +1265,43 @@ init_group_colocated_nodes(const pe_resource_t *rsc, const char *log_id,
+ * \internal
+ * \brief Initialize colocated node table for a non-group resource
+ *
+ * \param[in] rsc     Group resource being colocated with another resource
```
not group

kgaillot replied:
whoops

nrwahl2 commented on lib/pacemaker/pcmk_sched_colocation.c:
```diff
@@ -1265,6 +1265,43 @@ init_group_colocated_nodes(const pe_resource_t *rsc, const char *log_id,
+ * \param[in] factor  Incorporate scores multiplied by this factor
+ * \param[in] flags   Bitmask of enum pcmk__coloc_select values
+ *
+ * \return Table of node scores initialized for colocation
```
ditto on maybe adding a free-ing note

nrwahl2 commented on lib/pacemaker/pcmk_sched_colocation.c:
```diff
@@ -1220,6 +1220,51 @@ add_node_scores_matching_attr(GHashTable *nodes, const pe_resource_t *rsc,
+
+    if (*nodes == NULL) {
+        // Only cmp_resources() passes a NULL nodes table
+        member = pe__last_group_member(rsc);
```
This is probably correct and it matches the existing logic, but the reasoning for choosing the last resource instead of the first in this case isn't immediately obvious to me. I haven't had time to do a deeper dive yet.

Your commit message in an earlier commit says that `pcmk__group_assign()` adds the group's colocations to the last member. But `cmp_resources()` appears to run before the assign function, so assign hasn't been called yet.

kgaillot replied:
I hadn't gotten as far as dealing with that yet. I suspect it has issues. The PR was going to be scary long if I kept going :)

nrwahl2 commented on lib/pacemaker/pcmk_sched_colocation.c:
```diff
@@ -130,6 +130,34 @@ cmp_primary_priority(gconstpointer a, gconstpointer b)
+ * \brief Add a "this with" colocation constraint to a resource
+ *
+ * \param[in,out] rsc         Resource to add colocation to
+ * \param[in]     colocation  Colocation constraint to add
```
If we say "to \p rsc" in the "with this" function, we might as well here

nrwahl2 commented on lib/pacemaker/pcmk_sched_group.c:
```diff
@@ -61,20 +64,23 @@ expand_group_colocations(pe_resource_t *rsc)
         }
     }
 
+    g_list_free(rsc->rsc_cons);
```
This only frees the list and doesn't touch the elements, right? AFAICT only `g_list_free_full()` frees the elements, but the doc says "The freed elements are returned to the slice allocator." Not sure off hand what that means.

https://docs.gtk.org/glib/type_func.List.free.html

kgaillot replied:
Correct, it's a little obscure here: pe_working_set_t has a complete list of all colocation constraints in the cluster, then each resource has the constraints affecting it added to its rsc_cons or rsc_cons_lhs. Therefore, we free only the lists, not the constraints themselves, when freeing rsc_cons and rsc_cons_lhs; the constraints are freed when the pe_working_set_t is freed.

nrwahl2 reviewed (commented):
Looks great

kgaillot commented on cts/cts-scheduler.in:
```diff
@@ -235,6 +235,7 @@ TESTS = [
         [ "anti-colocation-promoted", "Organize order of actions for promoted resources in anti-colocations" ],
         [ "anti-colocation-unpromoted", "Organize order of actions for unpromoted resources in anti-colocations" ],
         [ "group-anticolocation", "Group with failed last member anti-colocated with another group" ],
+        [ "group-colocation-failure", "Group with failed last member colocated with another group" ],
```
Good point, it is specific to single-member groups

kgaillot commented on cts/scheduler/scores/colocation-priority-group.scores:
```diff
@@ -0,0 +1,19 @@
+pcmk__primitive_assign: Fencing allocation score on node1: 0
+pcmk__primitive_assign: Fencing allocation score on node2: 0
+pcmk__primitive_assign: member1a allocation score on node1: 0
+pcmk__primitive_assign: member1a allocation score on node2: -4000
```
Yeah, in pcmk__apply_coloc_to_weights(), if the primary isn't running, the negative of the score is applied to all the dependent's allowed nodes. Since it's the same for all nodes, presumably it would matter only if INFINITY. (Anti-colocations have already been filtered out by that point.)

nrwahl2 replied:
Ah okay, so it sounds like the 0 is from 4000 + (-4000) then too. (crm_simulate -VVVVVV probably would have told me that)

kgaillot commented on cts/cts-scheduler.in:
```diff
@@ -242,6 +242,7 @@ TESTS = [
         [ "coloc-dependee-should-move", "Group colocation outweighs stickiness" ],
         [ "colocation-influence", "Respect colocation influence" ],
         [ "colocation-priority-group", "Apply group colocations in order of primary priority" ],
+        [ "colocation-vs-stickiness", "Group stickiness outweighs anti-colocation score" ],
```
The test description isn't intended to mean that stickiness always outweighs anti-colocations, but that stickiness should win with this particular input because its score is higher.

Clarifying the docs is a good idea

kgaillot commented on lib/pacemaker/pcmk_sched_colocation.c:
```diff
@@ -1220,6 +1220,51 @@ add_node_scores_matching_attr(GHashTable *nodes, const pe_resource_t *rsc,
+ * \param[in] factor  Incorporate scores multiplied by this factor
+ * \param[in] flags   Bitmask of enum pcmk__coloc_select values
+ *
+ * \return Table of node scores initialized for colocation
```
Might as well

kgaillot commented:
Updated per review:

* Description for group-colocation-failure test updated
* Added documentation commit on top
* Updated doxygen blocks for init\_\*\_colocated_nodes() and pcmk\_\_add\_this\_with()

nrwahl2 commented on doc/sphinx/Pacemaker_Explained/constraints.rst:
```diff
@@ -585,15 +585,13 @@ Advisory Placement
+**INFINITY**, the cluster will try to accommodate your wishes, but may ignore
+them if other factors outweigh the colocation score. Those factors might
+include other constraints, resource stickiness, failure thresholds, whether
+other resources would be prevented from being active, etc.
```
Much clearer :)
